### ISSUE TITLE
Incorrect strain loss?

### ISSUE THREAD

**Issue by opendeeple** [isantesteban/snug#12]
Hi, Your strain loss function implementation is looking it does not match with formulas on the paper.
Formula:
![image](https://user-images.githubusercontent.com/74002741/172900174-25dba38a-346e-4f89-b213-61c8fad4aec3.png)
![image](https://user-images.githubusercontent.com/74002741/172900272-8784c34b-7c27-4a13-9a43-e4d093e02b14.png)
Strain loss function:
```py
def stretching_energy(v, cloth, return_average=True):
  batch_size = tf.cast(tf.shape(v)[0], v.dtype)
  triangles = gather_triangles(v, cloth.faces)

  Dm_inv = tf.repeat([cloth.Dm_inv], tf.shape(v)[0], axis=0)

  F = deformation_gradient(triangles, Dm_inv)
  G = green_strain_tensor(F)

  mat = cloth.material
  I = tf.eye(2, batch_shape=tf.shape(G)[:2], dtype=G.dtype)

  S = mat.lame_mu * G + 0.5 * mat.lame_lambda * tf.linalg.trace(G)[:, :, tf.newaxis, tf.newaxis] * I
  energy_density = tf.linalg.trace(tf.transpose(S, [0, 1, 3, 2]) @ G)
  energy = cloth.f_area[tf.newaxis] * mat.thickness * energy_density

  if return_average:
    return tf.reduce_sum(energy) / batch_size
  return tf.reduce_sum(energy, axis=-1)
```
For test, I have implement strain loss function based on snug paper.
```py
def stretching_energy(v, cloth, mu=11.1, lam=20.9, return_average=True):
  batch_size = tf.cast(tf.shape(v)[0], v.dtype)
  triangles = gather_triangles(v, cloth.faces)

  Dm_inv = tf.repeat([cloth.Dm_inv], tf.shape(v)[0], axis=0)

  F = deformation_gradient(triangles, Dm_inv)
  G = green_strain_tensor(F)

  mat = cloth.material

  Equation8 = 0.5 * lam * tf.pow(tf.linalg.trace(G), 2) + mu * tf.linalg.trace(tf.pow(G, 2))
  Equation9 = cloth.f_area[tf.newaxis] * mat.thickness * Equation8

  if return_average:
    return tf.reduce_sum(Equation9) / batch_size
  return tf.reduce_sum(Equation9, axis=-1)
```
Results:
```
YOURS: 41.80565 
OURS:   0.017105
```
I want to ask, How is correct way about strain loss?
Thanks for your attention!

**Comment by liren2515** [comment]
@[opendeeple](https://github.com/opendeeple)
I think in your implementation
```
  Equation8 = 0.5 * lam * tf.pow(tf.linalg.trace(G), 2) + mu * tf.linalg.trace(tf.pow(G, 2))
```
the last `tf.pow(G, 2)` should be `tf.transpose(G, [0, 1, 3, 2]*G`, since $tr(G^2)$ actually means $tr(G^TG)$ if I understand correctly.

**Comment by opendeeple** [comment]
@liren2515
Thanks for your help?

How is $tr(G)^2$, should it be `tf.pow(tf.linalg.trace(G), 2)` or is it incorrect?

Thanks!!!

**Comment by liren2515** [comment]
> @liren2515 Thanks for your help?
> 
> How is tr(G)2, should it be `tf.pow(tf.linalg.trace(G), 2)` or is it incorrect?
> 
> Thanks!!!

@opendeeple I think this is correct.

**Comment by liren2515** [comment]
By the way, in your another issue about training pipeline, it seems you succeed in replicating the training, right? Is the image you show generated by the model trained from scratch? If yes, do you mind sharing your script for the training, or more details about it?

**Comment by opendeeple** [comment]
> By the way, in your another issue about training pipeline, it seems you succeed in replicating the training, right? Is the image you show generated by the model trained from scratch? If yes, do you mind sharing your script for the training, or more details about it?

@liren2515 Sorry, our team did not accept to share codes to public! It is a team work, and I can not make it as public.

**Comment by liren2515** [comment]
@opendeeple OK, I understand. But do you mind telling if you use some initialization strategy for the network and if you use the initial hidden states $h_0=0$ for the training or random values sampled from the gaussian distribution $\mathcal{N}(0, 0.1)$ as described in the paper. When I use the random values, it seems hard to train.

**Comment by opendeeple** [comment]
I have updated our strain loss function, but loss value does not match with native strain loss function.
```py
def stretching_energy(v, cloth, mu=11.1, lam=20.9, return_average=True):
  batch_size = tf.cast(tf.shape(v)[0], v.dtype)
  triangles = gather_triangles(v, cloth.faces)

  Dm_inv = tf.repeat([cloth.Dm_inv], tf.shape(v)[0], axis=0)

  F = deformation_gradient(triangles, Dm_inv)
  G = green_strain_tensor(F)

  mat = cloth.material

  Equation8 = 0.5 * lam * tf.pow(tf.linalg.trace(G), 2) + mu * tf.linalg.trace(tf.transpose(G, [0, 1, 3, 2]) @ G)
  Equation9 = cloth.f_area[tf.newaxis] * mat.thickness * Equation8

  if return_average:
    return tf.reduce_sum(Equation9) / batch_size
  return tf.reduce_sum(Equation9, axis=-1)
```
Loss values:
```
SNUG native untrained strain loss function value: 51.473682
OUR untrained strain loss function value: 0.024238078
```
I want to ask, How is correct way about strain loss?
Thanks!!!

**Comment by opendeeple** [comment]
> do you mind telling if you use some initialization strategy for the network and if you use the initial hidden states.

In run_snug.py I have used hidden states with $h_0=0$ initialization, but in train_snug.py (It is our training script) I have used hidden states with $\mathcal{N}(0, 0.1)$ initialization as described in the paper.

I got something from [this unofficial repo](https://github.com/zihui-debug/snug-recurrence), I hope this helps.

**Comment by liren2515** [comment]
> > do you mind telling if you use some initialization strategy for the network and if you use the initial hidden states.
> 
> In run_snug.py I have used hidden states with h0=0 initialization, but in train_snug.py (It is our training script) I have used hidden states with N(0,0.1) initialization as described in the paper.
> 
> I got something from [this unofficial repo](https://github.com/zihui-debug/snug-recurrence), I hope this helps.

Thanks for your information!

**Comment by hsyntemiz** [comment]
> I got something from [this unofficial repo](https://github.com/zihui-debug/snug-recurrence), I hope this helps.

Thank you for sharing unofficial repo. I think you already try it. Does 'unofficial repo' seem to replicate the results? It is very helpful, and helps me to clarify many points. But, I am still suspicious about training.

**Comment by fishfishson** [comment]
> I have updated our strain loss function, but loss value does not match with native strain loss function.
> 
> ```python
> def stretching_energy(v, cloth, mu=11.1, lam=20.9, return_average=True):
>   batch_size = tf.cast(tf.shape(v)[0], v.dtype)
>   triangles = gather_triangles(v, cloth.faces)
> 
>   Dm_inv = tf.repeat([cloth.Dm_inv], tf.shape(v)[0], axis=0)
> 
>   F = deformation_gradient(triangles, Dm_inv)
>   G = green_strain_tensor(F)
> 
>   mat = cloth.material
> 
>   Equation8 = 0.5 * lam * tf.pow(tf.linalg.trace(G), 2) + mu * tf.linalg.trace(tf.transpose(G, [0, 1, 3, 2]) @ G)
>   Equation9 = cloth.f_area[tf.newaxis] * mat.thickness * Equation8
> 
>   if return_average:
>     return tf.reduce_sum(Equation9) / batch_size
>   return tf.reduce_sum(Equation9, axis=-1)
> ```
> 
> Loss values:
> 
> ```
> SNUG native untrained strain loss function value: 51.473682
> OUR untrained strain loss function value: 0.024238078
> ```
> 
> I want to ask, How is correct way about strain loss? Thanks!!!

You multiply thickness twice! Please see the errata of this repo.

**Comment by opendeeple** [comment]
@fishfishson Thanks for your attention!
On official implementation, multiplying *area <sub><sup>x</sub></sup> thickness* to calculate V value on the Equation 9.
I have used this method to one. Can you tell more if I understand incorrectly?

**Comment by fishfishson** [comment]
> @fishfishson Thanks for your attention! On official implementation, multiplying _area x thickness_ to calculate V value on the Equation 9. I have used this method to one. Can you tell more if I understand incorrectly?

Do you mean set _area x thickness_ to 1? According to the errata, the correct values of the lamé coefficients are μ=2.36e4 and λ=4.44e4. I think you could get consistent results with the author if you parse them to your _stretching_energy_ function.
